File: frontend/src/components/Risk/cast.ts
// Client-side display info for the AI generals. The backend (api/services/risk/cast.go)
// is the source of truth for strategy weights; this file only mirrors the
// fields needed to render avatars + blurbs in the lobby and HUD.

export interface General {
  id: string;
  name: string;
  title: string;
  blurb: string;
  color: string; // Mantine palette key (display tint, not the in-game player color)
  emoji: string; // fallback avatar when dotLottie cannot render
  avatarSrc: string;
  avatarAlt: string;
  artDirection: string[];
}

export const GENERALS: General[] = [
  {
    id: 'maxine-voltage',
    name: 'Maxine Voltage',
    title: 'Field Marshal',
    blurb: 'A neon-soaked cavalry commander who treats the world like a synth solo — loud, fast, impossible to ignore.',
    color: 'neonPink',
    emoji: '⚡',
    avatarSrc: '/risk-avatars/maxine-voltage.lottie',
    avatarAlt: 'Animated neon-pink avatar for Maxine Voltage',
    artDirection: ['neon cavalry', 'lightning insignia', 'hot pink command glow'],
  },
  {
    id: 'general-static',
    name: 'General Static',
    title: 'Commander',
    blurb: 'A patient tactician whose silences are louder than her artillery. She finishes continents others abandon.',
    color: 'electricBlue',
    emoji: '📺',
    avatarSrc: '/risk-avatars/general-static.lottie',
    avatarAlt: 'Animated electric-blue avatar for General Static',
    artDirection: ['crt visor', 'blue scanlines', 'quiet tactician'],
  },
  {
    id: 'vice-admiral-vector',
    name: 'Vice-Admiral Vector',
    title: 'Vice-Admiral',
    blurb: 'Naval strategist with a vendetta against any continent that touches the sea. Especially Australia.',
    color: 'neonGreen',
    emoji: '⚓',
    avatarSrc: '/risk-avatars/vice-admiral-vector.lottie',
    avatarAlt: 'Animated green admiral avatar for Vice-Admiral Vector',
    artDirection: ['naval cap', 'anchor insignia', 'green vector glow'],
  },
  {
    id: 'commodore-cassette',
    name: 'Commodore Cassette',
    title: 'Commodore',
    blurb: 'Rewinds, replays, and refuses to flip sides. Predictable but relentless.',
    color: 'hotYellow',
    emoji: '📼',
    avatarSrc: '/risk-avatars/commodore-cassette.lottie',
    avatarAlt: 'Animated yellow avatar for Commodore Cassette',
    artDirection: ['cassette badge', 'tape-loop energy', 'yellow deck lights'],
  },
  {
    id: 'captain-coral',
    name: 'Captain Coral',
    title: 'Captain',
    blurb: 'Coastal raider who hops across oceans. Reads the wrap-around adjacencies like nobody else.',
    color: 'synthPurple',
    emoji: '🐚',
    avatarSrc: '/risk-avatars/captain-coral.lottie',
    avatarAlt: 'Animated purple coastal avatar for Captain Coral',
    artDirection: ['shell insignia', 'coral raider', 'violet sea glow'],
  },
  {
    id: 'field-marshal-neon',
    name: 'Field Marshal Neon',
    title: 'Field Marshal',
    blurb: 'Old-school continental conquest doctrine in a chrome-pink uniform.',
    color: 'neonPink',
    emoji: '🎖',
    avatarSrc: '/risk-avatars/field-marshal-neon.lottie',
    avatarAlt: 'Animated chrome-pink avatar for Field Marshal Neon',
    artDirection: ['field marshal cap', 'medal insignia', 'old-school neon doctrine'],
  },
  {
    id: 'colonel-chrome',
    name: 'Colonel Chrome',
    title: 'Colonel',
    blurb: 'Polished to a mirror. Plays a textbook board, opens with Asia, dares you to stop him.',
    color: 'electricBlue',
    emoji: '🔱',
    avatarSrc: '/risk-avatars/colonel-chrome.lottie',
    avatarAlt: 'Animated chrome-blue avatar for Colonel Chrome',
    artDirection: ['chrome visor', 'trident insignia', 'mirror-polished uniform'],
  },
  {
    id: 'lieutenant-laser',
    name: 'Lt. Laser',
    title: 'Lieutenant',
    blurb: 'Junior officer with a bright pink targeting grid. Over-commits, but lands the first kill.',
    color: 'hotYellow',
    emoji: '🎯',
    avatarSrc: '/risk-avatars/lieutenant-laser.lottie',
    avatarAlt: 'Animated yellow target-grid avatar for Lieutenant Laser',
    artDirection: ['targeting grid', 'junior laser officer', 'hot yellow lock-on'],
  },
];

export function generalById(id: string | undefined | null): General | undefined {
  if (!id) return undefined;
  return GENERALS.find((g) => g.id === id);
}

File: api/services/risk/cast.go
package risk

// General is one of the synthwave-themed AI opponents. Difficulty + Bias
// together shape the AI's strategy weights at runtime; the persona is also
// surfaced on the frontend for character art + flavor.
type General struct {
	ID    string
	Name  string
	Title string
	Blurb string
	Color string // Mantine palette key
	Bias  Bias
}

// Bias is a per-general personality vector layered on top of the difficulty
// knob. All values are 0..1.
type Bias struct {
	Aggression      float64 // higher → attacks more often and at lower odds
	ContinentFocus  float64 // higher → harder to dislodge from a continent goal
	RiskTolerance   float64 // higher → accepts marginal EV attacks
	BorderHardening float64 // higher → stacks borders, leaves interiors thin
}

// Generals is the synthwave-themed Risk cast. Eight entries — enough variety
// for any 2-6 player game with no repeats. Each general's color is a hint;
// the engine still assigns a unique board color from the player palette.
var Generals = []General{
	{
		ID: "maxine-voltage", Name: "Maxine Voltage", Title: "Field Marshal",
		Blurb: "A neon-soaked cavalry commander who treats the world like a synth solo — loud, fast, and impossible to ignore.",
		Color: "neonPink",
		Bias:  Bias{Aggression: 0.85, ContinentFocus: 0.45, RiskTolerance: 0.7, BorderHardening: 0.4},
	},
	{
		ID: "general-static", Name: "General Static", Title: "Commander",
		Blurb: "A patient tactician whose silences are louder than her artillery. She finishes continents others abandon.",
		Color: "electricBlue",
		Bias:  Bias{Aggression: 0.35, ContinentFocus: 0.85, RiskTolerance: 0.4, BorderHardening: 0.75},
	},
	{
		ID: "vice-admiral-vector", Name: "Vice-Admiral Vector", Title: "Vice-Admiral",
		Blurb: "Naval strategist with a vendetta against any continent that touches the sea. Especially Australia.",
		Color: "neonGreen",
		Bias:  Bias{Aggression: 0.55, ContinentFocus: 0.75, RiskTolerance: 0.55, BorderHardening: 0.5},
	},
	{
		ID: "commodore-cassette", Name: "Commodore Cassette", Title: "Commodore",
		Blurb: "Rewinds, replays, and refuses to flip sides. Predictable but relentless — once she picks a target, she will not stop.",
		Color: "hotYellow",
		Bias:  Bias{Aggression: 0.5, ContinentFocus: 0.65, RiskTolerance: 0.5, BorderHardening: 0.6},
	},
	{
		ID: "captain-coral", Name: "Captain Coral", Title: "Captain",
		Blurb: "Coastal raider who hops across oceans. Reads the wrap-around adjacencies like nobody else.",
		Color: "synthPurple",
		Bias:  Bias{Aggression: 0.75, ContinentFocus: 0.3, RiskTolerance: 0.8, BorderHardening: 0.3},
	},
	{
		ID: "field-marshal-neon", Name: "Field Marshal Neon", Title: "Field Marshal",
		Blurb: "Old-school continental conquest doctrine in a chrome-pink uniform. Wins continents one by one and never gives them back.",
		Color: "neonPink",
		Bias:  Bias{Aggression: 0.5, ContinentFocus: 0.9, RiskTolerance: 0.45, BorderHardening: 0.85},
	},
	{
		ID: "colonel-chrome", Name: "Colonel Chrome", Title: "Colonel",
		Blurb: "Polished to a mirror. Plays a textbook board with one twist: he opens with Asia and dares you to stop him.",
		Color: "electricBlue",
		Bias:  Bias{Aggression: 0.6, ContinentFocus: 0.7, RiskTolerance: 0.6, BorderHardening: 0.55},
	},
	{
		ID: "lieutenant-laser", Name: "Lt. Laser", Title: "Lieutenant",
		Blurb: "Junior officer with a bright pink targeting grid. Tends to over-commit, but lands the first kill of every game.",
		Color: "hotYellow",
		Bias:  Bias{Aggression: 0.95, ContinentFocus: 0.25, RiskTolerance: 0.85, BorderHardening: 0.25},
	},
}

// GeneralByID returns the persona definition or false.
func GeneralByID(id string) (General, bool) {
	for _, g := range Generals {
		if g.ID == id {
			return g, true
		}
	}
	return General{}, false
}

// PlayerColors is the cycling palette the engine assigns to slots so each
// player has a distinct, theme-matched color regardless of which generals
// were selected.
var PlayerColors = []string{
	"neonPink", "electricBlue", "neonGreen", "hotYellow", "synthPurple", "iceCyan",
}

// PickGenerals selects n distinct AI generals deterministically given a seed
// (so a "Restart" with the same settings can produce the same opponents if
// desired). For variety the order is rotated through the static slice.
func PickGenerals(n, rotation int) []General {
	if n <= 0 {
		return nil
	}
	if n > len(Generals) {
		n = len(Generals)
	}
	out := make([]General, 0, n)
	for i := 0; i < n; i++ {
		out = append(out, Generals[(i+rotation)%len(Generals)])
	}
	return out
}
